tasks/model.go

```diff
@@ -188,11 +135,126 @@ func executeStage(stage string, updateStage UpdateStage, steps []step) error {
 		if err != nil {
 			return err
 		}
-		AddLog(&stageDetails, step.stepSuccess)
-		err = updateStage(stage, &stageDetails)
+		stageDetails = AddLog(stageDetails, step.stepSuccess)
+		err = updateStage(stage, stageDetails)
 		if err != nil {
 			return nil
 		}
 	}
 	return nil
 }
+
+func (sdp *_StageDetails__Prototype) Of(desc, expected string) *_StageDetails {
+	sd := _StageDetails{
+		Description:      _String__Maybe{m: schema.Maybe_Value, v: &_String{desc}},
+		ExpectedDuration: _String__Maybe{m: schema.Maybe_Value, v: &_String{expected}},
+		Logs:             _List_Logs{[]_Logs{}},
+		UpdatedAt:        _Time__Maybe{m: schema.Maybe_Value, v: &_Time{x: time.Now().UnixNano()}},
+	}
+	return &sd
+}
+
+// WithLog makes a copy of the stage details with an additional log appended.
+func (sd *_StageDetails) WithLog(log string) *_StageDetails {
+	nl := _List_Logs{
+		x: append(sd.Logs.x, _Logs{
+			Log:       _String{log},
+			UpdatedAt: _Time{x: time.Now().UnixNano()},
+		}),
+	}
+	n := _StageDetails{
+		Description:      sd.Description,
+		ExpectedDuration: sd.ExpectedDuration,
+		Logs:             nl,
+		UpdatedAt:        _Time__Maybe{m: schema.Maybe_Value, v: &_Time{x: time.Now().UnixNano()}},
+	}
+	return &n
+}
+
+func (t *_Time) Time() time.Time {
+	return time.Unix(0, t.x)
+}
+
+func (tp *_Task__Prototype) New(r RetrievalTask, s StorageTask) Task {
+	t := _Task{
+		UUID:                _String{uuid.New().String()},
+		Status:              *Available,
+		WorkedBy:            _String__Maybe{m: schema.Maybe_Value, v: &_String{""}},
+		Stage:               _String{""},
+		CurrentStageDetails: _StageDetails__Maybe{m: schema.Maybe_Absent},
+		StartedAt:           _Time__Maybe{m: schema.Maybe_Absent},
+		RetrievalTask:       _RetrievalTask__Maybe{m: schema.Maybe_Absent},
+		StorageTask:         _StorageTask__Maybe{m: schema.Maybe_Absent},
+	}
+	if r != nil {
+		t.RetrievalTask.m = schema.Maybe_Value
+		t.RetrievalTask.v = r
+	}
+	if s != nil {
+		t.StorageTask.m = schema.Maybe_Value
+		t.StorageTask.v = s
+	}
+	return &t
+}
+
+func (t *_Task) Assign(worker string, status Status) {
```

Comment:
I feel similarly uneasy about all these extra methods, but I guess they help manipulate and use the structs since the codegen types are pretty hard to use directly.